lib/fetch-motif-doc.js

```diff
@@ -0,0 +1,145 @@
+import _ from "lodash";
+import filterOverlappingPeaks from "./filter-overlapping-peaks";
+
+/**
+ * Return a list of motifs. Each motifs contain info about pwm id, href, targets, files, datasets, peak start and end,
+ * strand, biosamples, description and data matrix that will use for draw DNA logo.
+ * @param {object} request to fetch data for data matrix
+ * @param {Array} data contain motifs info
+ * @returns Return a list of motifs.
+ */
+export default async function fetchMotifDoc(request, data) {
+  const motifsData = filterOverlappingPeaks(
+    data.filter((d) => d.method === "footprints" || d.method === "PWMs")
+  );
+  // Filter results to find ones with PWM data
+  const pwmLinkListFull = motifsData.filter(
+    (d) =>
+      d.documents &&
+      d.documents[0] &&
+      d.documents[0].document_type === "position weight matrix"
+  );
+  // Find all pwms that have both matching document and matching target(s)
+  // Group pwms based on these two properties
+  // Result is object with keys that are "[document link]#[target list linked by '-'s]"
+  const groupedMotifsData = _.groupBy(
+    pwmLinkListFull,
+    (dataset) =>
+      `${dataset.documents[0]["@id"]}#${dataset.targets.sort().join("-")}`
+  );
+  // Flatten group to create an array of pwms
+  // Properties that are identical across the group (for instance: pwm document, strand, and targets) are collapsed
+  // Properties that are not the same across the group (for instance: biosamples, accessions, datasets) are merged into an array
+  // Also fetch pwm data for each motif from Encode
+  const motifsList = await Promise.all(
+    _.map(groupedMotifsData, async (group) => {
+      const pwm = group[0].documents[0]["@id"];
+      const href = group[0].documents[0].attachment.href;
+      const fileLink = `${pwm}${href}`;
+
+      const data = await request.getTextByUrl(fileLink);
+      const fileType = pwm.includes(".pwm") ? "pwm" : "jaspar";
+      const dataMatrix =
+        fileType === "pwm"
+          ? convertPwmTextToObj(data)
+          : convertJasparTextToObj(data);
+
+      return {
+        pwm,
+        href,
+        targets: group[0].targets.join(", "),
+        accessions: _.map(group, "file"),
+        datasets: _.map(group, "dataset"),
+        start: group[0].start,
+        end: group[0].end,
+        strand: group[0].strand,
+        biosamples: _.map(group, "biosample_ontology")
+          .map((d) => d.term_name)
+          .map((termName) => termName || ""),
+        description: group[0].description ? group[0].description : null,
+        data,
+        dataMatrix,
+      };
+    })
```

Comment:
Does `_.map()` have an advantage over `groupedMotifsData.map()`?

Reply:
hmmm, Didn't think about it when I just copied the original code over. Now when I look at it, I think they are similar but _.map produce shorter code and maybe a little bit easier to understatnd?